### DISCUSSION TITLE
VIPS installed with HEIF support can't load / save .heif files (Unsupported codec / compression)

### DISCUSSION THREAD

**Discussion by zackpi** [libvips/libvips#3173] [General]
# Describe the bug

I'm trying to build VIPS with HEIF support on Amazon Linux.
I am able to build and install it, and confirm that HEIF support is available.
Nonetheless when I try to operate on a .heif / .heic file, I get an "Unsupported codec" error on heifload, and a "Unsupported compression" error on heifsave.

# To Reproduce
### Use Image

[heif](https://filesamples.com/samples/image/heif/sample1.heif)
[jpeg](https://filesamples.com/samples/image/jpeg/sample_640×426.jpeg)

### Use Configuration
Build libx265, libde265, libheif, and libvips from source using the following script:

```bash
export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
DE265_VERSION=1.0.9
HEIF_VERSION=1.14.0
VIPS_VERSION=8.13.3

# compiling libx265
yum install -y mercurial
hg clone http://hg.videolan.org/x265
cd ./x265/build/linux && \
        cmake -G "Unix Makefiles" ../../source && \
        make -j6 && \
        make install && \
        ldconfig

# compiling libde265
curl -L https://github.com/strukturag/libde265/releases/download/v$DE265_VERSION/libde265-$DE265_VERSION.tar.gz | tar xz
cd libde265-$DE265_VERSION && \
        ./autogen.sh && \
        ./configure --disable-sherlock265 && \
        make -j6 && \
        make install

# compiling libheif
curl -L https://github.com/strukturag/libheif/releases/download/v$HEIF_VERSION/libheif-$HEIF_VERSION.tar.gz | tar xz
cd libheif-$HEIF_VERSION && \
        ./autogen.sh && \
        ./configure && \
        make -j6 && \
        make install

# compiling libvips
curl -L https://github.com/libvips/libvips/releases/download/v$VIPS_VERSION/vips-$VIPS_VERSION.tar.gz | tar xz
cd vips-$VIPS_VERSION && \
        ./configure
        make -j6 && \
        make install && \
        echo /usr/local/lib > /etc/ld.so.conf.d/libvips.conf && \
        ldconfig
```

During the script, `configure` for `libheif` shows:
```bash
...
configure: libde265 decoder: yes
configure: libx265 encoder: yes
configure: JPEG output: yes
...
```

and `configure` for `libvips` shows:
```bash
...
checking for libheif >= 1.3.0... yes
checking for heif_image_handle_get_raw_color_profile... yes
checking for heif_context_set_maximum_image_size_limit... yes
checking for libheif >= 1.7.0... yes
checking for struct heif_decoding_options.convert_hdr_to_8bit... yes
...
```

After running the script, I confirmed correct installation of VIPS with HEIF support:

```bash
> vips --vips-version
libvips 8.13.3-Tue Nov  1 09:09:54 UTC 2022

> vips --vips-config | grep "heif"
HEIC/AVIF load/save with libheif: yes (dynamic module: no)

> vips -l | grep "Heif"
        VipsForeignLoadHeif (heifload_base), load a HEIF image, priority=0
          VipsForeignLoadHeifFile (heifload), load a HEIF image (.heic, .heif, .avif), priority=0, is_a, get_flags, header, load
          VipsForeignLoadHeifBuffer (heifload_buffer), load a HEIF image, priority=0, is_a_buffer, get_flags, header, load
          VipsForeignLoadHeifSource (heifload_source), load a HEIF image, priority=0, is_a_source, get_flags, header, load
        VipsForeignSaveHeif (heifsave_base), save image in HEIF format, priority=0, rgba-only
          VipsForeignSaveHeifFile (heifsave), save image in HEIF format (.heic, .heif, .avif), priority=0, rgba-only
          VipsForeignSaveHeifBuffer (heifsave_buffer), save image in HEIF format (.heic, .heif), priority=0, rgba-only
          VipsForeignSaveHeifTarget (heifsave_target), save image in HEIF format (.heic, .heif), priority=0, rgba-only

```

### See error

By all accounts, it seems VIPS is installed with HEIF support. But when I try to load or save a .heif / .heic file, I get the following errors:
```bash
> vips copy sample1.heif out.jpg
sample1.heif: bad seek to 917718
heif: Unsupported feature: Unsupported codec (4.3000)

> vips copy sample_640×426.jpg out.heif 
heifsave: Unsupported compression
```

**Expected behavior**
A VIPS installation that claims to support HEIF load / save should be able to load and save HEIF files.

**Actual behavior**
VIPS throws an error saying HEIF load / save are not supported

**Screenshots**
n/a

**Environment**
 - OS: Amazon Linux
 - Vips: 8.13.3

**Additional context**

**Comment by lovell** [comment]
Does running `heif-convert sample1.heif out.jpg` and `heif-convert sample_640×426.jpg out.heif` also fail in a similar manner?

**Reply by zackpi** [reply]
Worse, I don't even have `heif-convert` installed.
Strangely, I do have the `man` page... 😵‍💫 

```bash
> heif-convert
-bash: heif-convert: command not found

> locate heif-convert
/usr/local/share/man/man1/heif-convert.1

> man heif-convert
# works; opens man page

> find -name "heif-convert"
# no results
```

Do you see anything wrong with the script that I used?
I am trying to install this on Amazon Linux, with the goal of using it as a custom build for Sharp.
I derived the script from the discussion in #1844.

**Comment by kleisauke** [comment]
> ```shell
> export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
> ```

Perhaps that should point to `/usr/local/lib64/pkgconfig` instead? What do these commands print?:
```shell
# HEIC
$ pkg-config libheif --variable=builtin_h265_encoder
$ heif-enc --list-encoders

# AVIF
$ pkg-config libheif --variable=builtin_avif_encoder
$ heif-enc -A --list-encoders
```

Alternatively, you don't need to compile this yourself on Amazon Linux 2, you can just use libvips (and the dynamic libheif module) from Remi's RPM repository, see: https://github.com/libvips/libvips/issues/1184#issuecomment-447973135.

<details>
  <summary>Optional modules</summary>

```shell
## Install optional modules
## See: https://github.com/remicollet/remirepo/issues/171

# HEIC / AVIF support via libheif
yum install --nogpgcheck https://mirrors.rpmfusion.org/free/el/rpmfusion-free-release-7.noarch.rpm
yum install vips-heif

# *Magick support via GraphicsMagick
yum install vips-magick-gm

# *Magick support via ImageMagick 6
yum install vips-magick-im6

# *Magick support via ImageMagick 7
yum install vips-magick-im7

# OpenSlide support
yum install vips-openslide

# PDF load support via Poppler
yum install vips-poppler
```
</details>

However, since Amazon Linux 2 is a derivative of RHEL 7 it ends up installing [libheif 1.4.0](https://git.remirepo.net/cgit/rpms/lib/libheif.git/tree/libheif.spec?id=53171b09fb723c025cb562be28313487c923c30b#n11), which is quite old already. You could try opening an issue at https://github.com/remicollet/remirepo/issues if that's a problem for you.

**Reply by zackpi** [reply]
okay, seems like libheif wasn't building properly and I didn't notice. The dependencies were being built to /usr/lib and it wasn't finding them.
Seems like it's all set now though. thanks!

**Reply by galdolber** [reply]
@zackpi I'm having the same issue, how did you fix it?

**Comment by remicollet** [comment]
@kleisauke FYI, using **libheif** from Fedora or EPEL-9 repository, you also need **libheif-freeworld** from [RPMfusion](https://rpmfusion.org/) (x265 HEVC encoder)
See https://github.com/remicollet/remirepo/issues/248
